File: 2477-minimum-fuel-cost-to-report-to-the-capital/2477-minimum-fuel-cost-to-report-to-the-capital.cpp
class Solution {
public:
    long long ans = 0;
    long long dfs(vector<vector<int>>& adj , int seats , int i , int par){
        long long people = 1;
        for(int e : adj[i]){
            if(e == par) continue;
            long long x = dfs(adj , seats , e , i);
            people += x;
            ans += (x + seats - 1) / seats;
        }
        return people;
    }
    long long minimumFuelCost(vector<vector<int>>& roads, int seats) {
        int n = roads.size();
        vector<vector<int>> adj(n + 1);
        for(int i = 0 ; i < n ;i++){
            int u = roads[i][0] , v = roads[i][1];
            adj[u].push_back(v);
            adj[v].push_back(u);
        }
        dfs(adj , seats , 0 , -1);
        return ans ;
    }
};

File: 2477-minimum-fuel-cost-to-report-to-the-capital/2477-minimum-fuel-cost-to-report-to-the-capital.java
class Solution {
    long ans = 0;
    long dfs(List<List<Integer>> adj , int seats, int node , int par){
        long p = 1;
        for(int i : adj.get(node)){
            if(i == par) continue;
            long x = dfs(adj , seats , i , node);
            p += x;
            ans += (x + seats - 1) / seats;
        }
        return p;
    }

    public long minimumFuelCost(int[][] roads, int seats) {
        int n = roads.length;
        List<List<Integer>> adj = new ArrayList<>(n + 1);
        for(int i = 0 ; i <= n ; i++){
            adj.add(new ArrayList<>());
        }
        for(int i = 0 ; i < n ;i++){
            int u = roads[i][0] , v = roads[i][1];
            adj.get(u).add(v);
            adj.get(v).add(u);
        }
        dfs(adj , seats , 0 , -1);
        return ans;
    }
}